### ISSUE TITLE
Turning off stax.Dropout for prediction

### ISSUE THREAD

**Issue by adamhaber** [google/jax#3321]
I'm trying to use stax.Dropout to regularise a simple model; I've followed the example from #170, with the necessary changes to fit the current APIs:
```python
from __future__ import print_function
import jax.numpy as np
from jax import jit, grad
from jax import random
from jax.experimental import stax, optimizers
from jax.experimental.stax import Dense, Relu, Dropout, LogSoftmax

init_fun, apply_fun = stax.serial(
    Dense(512), Relu, Dropout(0.4, mode='train'),
    Dense(512), Relu, Dropout(0.4, mode='train'),
    Dense(3), LogSoftmax
)

## Initialize parameters, not committing to a batch shape
in_shape = (-1, 28 * 28)
key = random.PRNGKey(0)
out_shape, net_params = init_fun(key, in_shape)

## Apply network to dummy inputs.

inputs = np.ones((10, 28 * 28))

key, subkey = random.split(key)
print(apply_fun(net_params, inputs, rng=subkey))

key, subkey = random.split(key)
print(apply_fun(net_params, inputs, rng=subkey))



## Run a training loop on dummy data

opt_init, opt_update, get_params = optimizers.momentum(step_size=1e-3, mass=0.9)

def loss(params, batch, key):
  inputs, targets = batch
  predictions = apply_fun(params, inputs, rng=key)
  return np.sum((predictions - targets) ** 2)

@jit
def step(i, opt_state, batch, key):
  params = get_params(opt_state)
  g = grad(loss)(params, batch, key)
  return opt_update(i, g, opt_state)

# Dummy input data stream
data_generator = ((np.zeros((10, 28 * 28)), np.zeros((10, 3)))
                  for _ in range(10))

# Optimize parameters in a loop
opt_state = opt_init(net_params)
for i in range(10):
  key, subkey = random.split(key)
  opt_state = step(i, opt_state, next(data_generator), subkey)
```
My question is - how can I turn off Dropout for prediction after training? I think I can construct a new serial object and use the post-training net_params without the all Dropouts' () tuples - is there a better way to do this?

**Comment by mattjj** [comment]
I think what you described is the right way to do it with stax: whether in training-mode or test-mode is baked into the network when it's built, as you've noted.

You could organize things like this:

```python
def make_net(mode: str):
  return stax.serial(
    Dense(512), Relu, Dropout(0.4, mode=mode),
    Dense(512), Relu, Dropout(0.4, mode=mode),
    Dense(3), LogSoftmax
  )

# for training
init_fun, apply_fun = make_net('train')

# for testing
_, apply_fun = make_net('test')
```

Alternatively you could make your own layer constructor that works differently (I _think_ this works, because IIRC you can stuff anything you want into the kwargs passed to an apply_fun...):

```python
def CoolDropout(rate):
  def init_fun(rng, input_shape):
    return input_shape, ()
  def apply_fun(params, inputs, **kwargs):
    mode = kwargs.get('mode', 'train')
    rng = kwargs.get('rng', None)
    if rng is None:
      msg = ("Dropout layer requires apply_fun to be called with a PRNG key "
             "argument. That is, instead of `apply_fun(params, inputs)`, call "
             "it like `apply_fun(params, inputs, rng)` where `rng` is a "
             "jax.random.PRNGKey value.")
      raise ValueError(msg)
    if mode == 'train':
      keep = random.bernoulli(rng, rate, inputs.shape)
      return jnp.where(keep, inputs / rate, 0)
    else:
      return inputs
  return init_fun, apply_fun
```

WDYT?

**Comment by mattjj** [comment]
By the way, for more fully-featured neural net libraries, you should check out [Flax](https://github.com/google/flax) and [Haiku](https://github.com/deepmind/dm-haiku), and [Trax](https://github.com/google/trax).

**Comment by mattjj** [comment]
I think this thread covered the right bases, so I'm going to close the issue. Please reopen or open new ones as needs arise!

**Comment by 8bitmp3** [comment]
> By the way, for more fully-featured neural net libraries, you should check out Flax, Haiku, and Trax.

@adamhaber There's also Objax (https://objax.readthedocs.io/en/latest/objax/nn.html#objax.nn.Dropout). 

@mattjj Do you know if there are plans to work on only 1 NN library at some point (other than DeepMind Haiku)? I've noticed there's a bit of confusion at the moment among some new users. Maybe centralizing the efforts around 1 NN library based on JAX can help with user adoption.
